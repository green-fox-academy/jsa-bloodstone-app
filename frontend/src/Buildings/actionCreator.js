import { SERVER_URL } from 'react-native-dotenv';

export const FETCH_BUILDINGS_REQUEST = 'fetchBuildingsRequest';
export const FETCH_BUILDINGS_SUCCESS = 'fetchBuildingsSuccess';
export const FETCH_BUILDINGS_FAILURE = 'fetchBuildingsFailure';
export const ADD_BUILDING_REQUEST = 'addBuildingRequest';
export const ADD_BUILDING_SUCCESS = 'addBuildingSuccess';
export const ADD_BUILDING_FAILURE = 'addBuildingFailure';

const URL = `http://${SERVER_URL}/kingdom/buildings`;

export function fetchBuildings(token) {
  return (dispatch) => {
    dispatch({ type: FETCH_BUILDINGS_REQUEST });
    return fetch(URL, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`,
      },
    })
      .then((response) => {
        if (response.status === 200) {
          return response.json();
        }
        throw new Error('An error has occurred, please try later!');
      })
      .then((response) => dispatch({ type: FETCH_BUILDINGS_SUCCESS, payload: response }))
      .catch((error) => dispatch({ type: FETCH_BUILDINGS_FAILURE, payload: error }));
  };
}

export function addBuilding(type, token) {
  return (dispatch) => {
    dispatch({ type: ADD_BUILDING_REQUEST });
    return fetch(URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ type }),
    })
      .then((response) => response.json())
      .then((response) => {
        if (response.status === 400) {
          return dispatch({ type: ADD_BUILDING_FAILURE, payload: response.message });
        }
        if (response.status === 200) {
          return dispatch({ type: ADD_BUILDING_SUCCESS, payload: response });
        }
        throw new Error(response.status);
      })
      .catch((error) => dispatch({ type: ADD_BUILDING_FAILURE, payload: error }));
  };
}
