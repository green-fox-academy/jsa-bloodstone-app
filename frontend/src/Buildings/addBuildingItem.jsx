import React from 'react';
import {
  View, Image, StyleSheet,
  TouchableHighlight, Text,
} from 'react-native';
import PropTypes from 'prop-types';
import Colors from '../common/colors';
import FontStyle from '../common/fonts';

const styles = StyleSheet.create({
  itemStyle: {
    alignItems: 'center',
    flexDirection: 'row',
    marginRight: 8,
  },
  iconStyle: {
    width: 32,
    height: 32,
  },
  textStyle: {
    marginLeft: 8,
    fontSize: FontStyle.fontSize,
    fontWeight: 'bold',
  },
  subText: {
    marginLeft: 8,
    fontSize: 10,
    color: Colors.orangeColor,
  },
});

function AddBuildingItem({
  icon, type, onPress, price,
}) {
  return (
    <TouchableHighlight underlayColor="transparent" onPress={onPress}>
      <View style={styles.itemStyle}>
        <Image style={styles.iconStyle} source={icon} />
        <View>
          <Text style={styles.textStyle}>{`Add ${type}`}</Text>
          <Text style={styles.subText}>{`-${price} golds`}</Text>
        </View>
      </View>
    </TouchableHighlight>
  );
}

AddBuildingItem.propTypes = {
  icon: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  onPress: PropTypes.func,
};

AddBuildingItem.defaultProps = {
  onPress: null,
};

export default AddBuildingItem;
