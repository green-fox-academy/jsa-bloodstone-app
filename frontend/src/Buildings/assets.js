import townhallIcon from '../../assets/buildings/townhall.png';
import academyIcon from '../../assets/buildings/academy.png';
import farmIcon from '../../assets/buildings/factory.png';
import mineIcon from '../../assets/buildings/mine.png';

const ICON_LIST = {
  Townhall: { type: 'Townhall', icon: townhallIcon },
  Academy: { type: 'Academy', icon: academyIcon },
  Farm: { type: 'Farm', icon: farmIcon },
  Mine: { type: 'Mine', icon: mineIcon },
};

export default function getIconImage(type) {
  return ICON_LIST[type].icon;
}
