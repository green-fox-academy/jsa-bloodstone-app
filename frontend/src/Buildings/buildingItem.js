import React from 'react';
import {
  View, TouchableHighlight,
  Image, Text, Dimensions,
  StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';

const buildingItemStyles = StyleSheet.create({
  itemStyle: {
    width: Dimensions.get('screen').width / 3,
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: 10,
    borderColor: 'black',
  },
  iconStyle: {
    width: 'auto',
    height: 'auto',
    maxWidth: 200,
    maxHeight: 200,
    minWidth: 70,
    minHeight: 70,
    aspectRatio: 1,
  },
  textStyle: {
    fontSize: 15,
    lineHeight: 20,
  },
});

function BuildingItem({
  type, level, onPress, getIconImage,
}) {
  return (
    <View style={buildingItemStyles.itemStyle}>
      <TouchableHighlight underlayColor="#0000" onPress={onPress}>
        <Image style={buildingItemStyles.iconStyle} source={getIconImage(type)} />
      </TouchableHighlight>
      <Text style={buildingItemStyles.textStyle}>{type}</Text>
      <Text style={buildingItemStyles.textStyle}>{`Level ${level}`}</Text>
    </View>
  );
}

BuildingItem.propTypes = {
  type: PropTypes.string.isRequired,
  level: PropTypes.number.isRequired,
  onPress: PropTypes.func,
  getIconImage: PropTypes.func,
};

BuildingItem.defaultProps = {
  onPress: null,
  getIconImage: null,
};

export default BuildingItem;
