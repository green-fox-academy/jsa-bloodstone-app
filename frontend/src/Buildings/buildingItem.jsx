import React from 'react';
import {
  View, TouchableHighlight,
  Image, Text,
  StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';
import Colors from '../common/colors';
import getIconImage from './assets';

const styles = StyleSheet.create({
  itemContainer: {
    margin: 5,
    padding: 24,
    paddingBottom: 0,
    backgroundColor: '#ffffff66',
    borderRadius: 10,
    alignItems: 'center',
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: 'rgba(0,0,0,.12)',
    flexBasis: 120,
  },
  imageContainer: {
    position: 'relative',
  },
  iconStyle: {
    width: 64,
    height: 64,
  },
  levelContainer: {
    paddingVertical: 2,
    paddingHorizontal: 6,
    position: 'absolute',
    top: -8,
    right: -12,
    borderRadius: 10,
    backgroundColor: Colors.blueColor,
  },
  levelStyle: {
    fontSize: 10,
    fontWeight: 'bold',
    color: Colors.whiteColor,
  },
  titleStyle: {
    fontSize: 14,
    fontWeight: 'bold',
    color: Colors.blackColor,
  },
  textContainer: {
    paddingVertical: 7,
    marginHorizontal: -4,
    alignItems: 'center',
  },
});

function BuildingItem({
  type, level, onPress,
}) {
  return (
    <TouchableHighlight
      underlayColor="transparent"
      onPress={onPress}
      style={{ flexBasis: '33%' }}
    >
      <View style={styles.itemContainer}>
        <View style={styles.imageContainer}>
          <Image style={styles.iconStyle} source={getIconImage(type)} />
          <View style={styles.textContainer}>
            <Text style={styles.titleStyle}>{type}</Text>
          </View>
          <View style={styles.levelContainer}>
            <Text style={styles.levelStyle}>{`Lv. ${level}`}</Text>
          </View>
        </View>
      </View>
    </TouchableHighlight>
  );
}

BuildingItem.propTypes = {
  type: PropTypes.string.isRequired,
  level: PropTypes.number.isRequired,
  onPress: PropTypes.func,
};

BuildingItem.defaultProps = {
  onPress: null,
};

export default BuildingItem;
