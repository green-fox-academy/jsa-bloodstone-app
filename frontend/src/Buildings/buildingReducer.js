import {
  FETCH_BUILDINGS_REQUEST,
  FETCH_BUILDINGS_SUCCESS,
  FETCH_BUILDINGS_FAILURE,
  ADD_BUILDING_REQUEST,
  ADD_BUILDING_SUCCESS,
  ADD_BUILDING_FAILURE,
} from './actionCreator';

const initialState = {
  listOfBuildings: [],
  buildingPrice: [],
  isLoading: false,
  error: undefined,
};

export default function buildings(state = initialState, action) {
  switch (action.type) {
    case FETCH_BUILDINGS_REQUEST:
    case ADD_BUILDING_REQUEST:
      return {
        ...state,
        isLoading: true,
      };
    case FETCH_BUILDINGS_SUCCESS:
      return {
        ...state,
        listOfBuildings: action.payload.buildings,
        buildingPrice: action.payload.priceOfBuilding,
        isLoading: false,
        error: undefined,
      };
    case ADD_BUILDING_SUCCESS:
      return {
        ...state,
        listOfBuildings: [...state.listOfBuildings, action.payload.newBuilding],
        isLoading: false,
        error: undefined,
      };
    case FETCH_BUILDINGS_FAILURE:
    case ADD_BUILDING_FAILURE:
      return {
        ...state,
        isLoading: false,
        error: action.payload,
      };
    default:
      return state;
  }
}
