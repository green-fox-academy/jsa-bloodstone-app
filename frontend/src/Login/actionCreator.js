export const LOGIN = 'login';
export const FORGOT_PASSWORD = 'forgotPassword';

export function login() {
  return {
    type: LOGIN,
  };
}

export function forgotPassword(input) {
  return {
    type: FORGOT_PASSWORD,
    payload: input,
  };
}
