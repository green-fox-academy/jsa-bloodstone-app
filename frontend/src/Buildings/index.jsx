/* eslint-disable no-underscore-dangle */
import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  View, ScrollView, StyleSheet, ActivityIndicator,
} from 'react-native';
import { Toast } from 'native-base';

import { fetchBuildings, addBuilding, ADD_BUILDING_FAILURE } from './actionCreator';

import addFarmIcon from '../../assets/buildings/addFarm.png';
import addMineIcon from '../../assets/buildings/addMine.png';

import Colors from '../common/colors';
import { CardView } from '../common/components';

import BuildingItem from './buildingItem';
import AddBuildingItem from './addBuildingItem';
import OneBuilding from '../OneBuilding';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 0,
    marginBottom: 0,
    paddingBottom: 52,
  },
  scrollViewContainer: {
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  addBuildingContainer: {
    bottom: 0,
    width: '100%',
    paddingVertical: 10,
    position: 'absolute',
    flexDirection: 'row',
    justifyContent: 'space-around',
    borderTopWidth: 1,
    borderTopColor: 'rgba(0,0,0,0.08)',
  },
  textStyle: {
    fontSize: 15,
    lineHeight: 20,
  },
});

const ADD_ICON_LIST = [
  { type: 'Farm', url: addFarmIcon },
  { type: 'Mine', url: addMineIcon },
];

function Buildings() {
  const dispatch = useDispatch();
  const buildings = useSelector((state) => state.buildings);
  const { listOfBuildings, buildingPrice, isLoading } = buildings;
  const { token } = useSelector((state) => state.auth);
  const [isModalVisible, setModalVisible] = useState(false);
  const [activeId, setActiveId] = useState('');

  useEffect(() => {
    dispatch(fetchBuildings(token));
  }, []);

  function handlePress(id) {
    setModalVisible(true);
    setActiveId(id);
  }

  async function addNewBuilding(type) {
    const { type: actionType, payload } = await dispatch(addBuilding(type, token));
    if (actionType === ADD_BUILDING_FAILURE) {
      Toast.show({
        type: 'warning',
        duration: 3000,
        text: payload,
        buttonText: 'Okay',
      });
    } else {
      Toast.show({
        type: 'success',
        duration: 3000,
        text: 'Building is succesfully added',
        buttonText: 'Okay',
      });
    }
  }

  if (isLoading) {
    return (
      <ActivityIndicator size="large" color={Colors.tealColor} />
    );
  }

  return (
    <View style={{ flex: 1 }}>
      <CardView style={styles.container}>
        <ScrollView
          bounces
          style={{ flex: 1 }}
          contentContainerStyle={styles.scrollViewContainer}
        >
          {listOfBuildings.map((building) => (
            <BuildingItem
              key={building._id}
              type={building.type}
              level={building.level}
              onPress={() => handlePress(building._id)}
            />
          ))}
        </ScrollView>
        <View style={styles.addBuildingContainer}>
          {ADD_ICON_LIST.map((addBuildingIcon, index) => (
            <AddBuildingItem
              key={addBuildingIcon.type}
              icon={addBuildingIcon.url}
              type={addBuildingIcon.type}
              onPress={() => addNewBuilding(addBuildingIcon.type)}
              price={buildingPrice[index]}
            />
          ))}
        </View>
      </CardView>
      {isModalVisible
        ? (
          <OneBuilding
            isVisible={isModalVisible}
            onClickClose={() => setModalVisible(false)}
            targetBuildingId={activeId}
          />
        )
        : null}
    </View>
  );
}

export default Buildings;
