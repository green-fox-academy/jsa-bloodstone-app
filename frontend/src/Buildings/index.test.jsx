import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { addBuildingSuccess } from './actionCreator';

import Buildings from '.';
import BuildingItem from './buildingItem';
import AddBuildingItem from './addBuildingItem';
import OneBuilding from '../OneBuilding';

jest.mock('react-redux');
jest.mock('./actionCreator');
jest.mock('./buildingItem', () => () => 'buildingItem');
jest.mock('./addBuildingItem', () => () => 'addBuildingItem');
jest.mock('../OneBuilding', () => () => 'oneBuilding');

const buildings = [
  {
    id: 1,
    type: 'Townhall',
    level: 1,
    hp: 1,
    started_at: 12345789,
    finished_at: 12399999,
  }, {
    id: 2,
    type: 'Academy',
    level: 1,
    hp: 1,
    started_at: 12345789,
    finished_at: 12399999,
  }, {
    id: 3,
    type: 'Farm',
    level: 1,
    hp: 1,
    started_at: 12345789,
    finished_at: 12399999,
  }, {
    id: 4,
    type: 'Mine',
    level: 1,
    hp: 1,
    started_at: 12345789,
    finished_at: 12399999,
  },
];

describe('<Buildings />', () => {
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Should opening oneBuilding after press BuildingItem', () => {
    const mockedState = {
      buildings: {
        listOfBuildings: buildings,
        isLoading: false,
        error: undefined,
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const testRenderer = renderer.create(<Buildings />);
    const testInstance = testRenderer.root;

    act(() => {
      testInstance.findAllByType(BuildingItem).forEach((node) => {
        node.props.onPress();
      });
    });

    expect(testInstance.findAllByType(OneBuilding).length).toBe(1);
  });

  it('Should add BuildingItems after press AddBuildingItem', () => {
    const mockedState = {
      buildings: {
        listOfBuildings: buildings,
        isLoading: false,
        error: undefined,
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));
    const mockDispatch = jest.fn(() => { });
    useDispatch.mockImplementation(() => mockDispatch);
    addBuildingSuccess.mockReturnValue('test_action');

    const testRenderer = renderer.create(<Buildings />);
    const testInstance = testRenderer.root;

    testInstance.findAllByType(AddBuildingItem).forEach((node) => {
      node.props.onPress();
    });
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith('test_action');
  });

  it('Should render buildings and match the snapshot', () => {
    const mockedState = {
      buildings: {
        listOfBuildings: buildings,
        isLoading: false,
        error: undefined,
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const tree = renderer.create(<Buildings />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render loading and match the snapshot', () => {
    const mockedState = {
      buildings: {
        listOfBuildings: buildings,
        isLoading: true,
        error: undefined,
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const tree = renderer.create(<Buildings />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render error and match the snapshot', () => {
    const mockedState = {
      buildings: {
        listOfBuildings: buildings,
        isLoading: true,
        error: new Error('test error'),
      },
    };
    useSelector.mockImplementation((func) => func(mockedState));

    const tree = renderer.create(<Buildings />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
