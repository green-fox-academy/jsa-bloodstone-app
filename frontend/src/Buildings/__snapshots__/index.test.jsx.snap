// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Buildings /> Should render buildings and match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <CardView
    style={
      Object {
        "flex": 1,
        "marginBottom": 0,
        "padding": 0,
        "paddingBottom": 52,
      }
    }
  >
    <ScrollViewMock
      bounces={true}
      contentContainerStyle={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Unknown
        level={1}
        onPress={[Function]}
        type="Townhall"
      />
      <Unknown
        level={1}
        onPress={[Function]}
        type="Academy"
      />
      <Unknown
        level={1}
        onPress={[Function]}
        type="Farm"
      />
      <Unknown
        level={1}
        onPress={[Function]}
        type="Mine"
      />
    </ScrollViewMock>
    <View
      style={
        Object {
          "borderTopColor": "rgba(0,0,0,0.08)",
          "borderTopWidth": 1,
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingVertical": 10,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <Unknown
        icon={1}
        onPress={[Function]}
        type="Farm"
      />
      <Unknown
        icon={1}
        onPress={[Function]}
        type="Mine"
      />
    </View>
  </CardView>
</View>
`;

exports[`<Buildings /> Should render error and match the snapshot 1`] = `
<ErrorPopup
  message="Oops, test error"
  title="System Error"
/>
`;

exports[`<Buildings /> Should render loading and match the snapshot 1`] = `
<ActivityIndicator
  animating={true}
  color="#10978b"
  hidesWhenStopped={true}
  size="large"
/>
`;
